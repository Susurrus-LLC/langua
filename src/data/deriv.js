export const defData = {
  words: 5,
  type: 'any',
  lexemes: [
    {
      lexeme: 'nzi',
      definition: 'tree'
    },
    {
      lexeme: 'ilu',
      definition: 'flower'
    },
    {
      lexeme: 'ru',
      definition: 'person'
    },
    {
      lexeme: 'kwərə',
      definition: 'goat'
    },
    {
      lexeme: 'lɛtsi',
      definition: 'sword'
    }
  ],
  derivations: [
    {
      derivation: '-vo',
      gloss: 'person',
      definition:
        'Creates a noun referring to a person associated with the base.'
    },
    {
      derivation: '-lu',
      gloss: 'place',
      definition: 'Creates a noun denoting the location of the base.'
    },
    {
      derivation: 'ña-',
      gloss: 'abstract',
      definition: 'Creates an abstract noun.'
    },
    {
      derivation: 'k-sa',
      gloss: 'collective',
      definition: 'Creates a noun referring to a collection of bases.'
    }
  ],
  results: undefined
}
