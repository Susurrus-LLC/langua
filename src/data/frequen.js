export const defData = {
  corpus:
    'Nzɔ de’áxa Tsənaxa, ño lo sɛtsa’ɔk, ño lo de’áxa u Kasaga, ño lo mɛya ayru Tol o Əbe kwə, sa nzɔ xa mpaskale: Steyə rabɛ lu ñire lu mvunə poxe o lu pɛlɔ ah lo mɛwɛ axe, maxə ɔra latsi əbɔwə səyo.',
  consonants:
    'b/d/f/g/k/kw/l/m/mp/mv/n/nt/nts/nz/ñ/ŋk/ŋkw/p/r/s/t/ts/v/w/x,h/y/’',
  vowels: 'a,á/ɛ,ɛ́/e,é/ə,ə́/i,í/ɔ,ɔ́/o,ó/u,ú',
  distinguishCase: false,
  analyzed: false,
  results: undefined,
  hovered: undefined,
  filterSeg: 'all',
  filterSyll: 'all',
  filterWords: 'all'
}
