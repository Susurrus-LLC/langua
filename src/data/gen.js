export const defData = {
  subpatterns: [
    {
      selected: 'V',
      subpattern: 'a/e/i/o/u'
    },
    {
      selected: 'C',
      subpattern: 'p/t/k/b/d/g/s/m/n/l/r'
    },
    {
      selected: 'N',
      subpattern: 'm/n'
    }
  ],
  pattern: 'V/CV/CVN/CVCV/CVCVN',
  words: 100,
  newline: false,
  filterdupes: false
}

export const vars = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z'
]
