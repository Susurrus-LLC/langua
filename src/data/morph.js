export const defData = {
  categories: [
    'V=ieaou',
    'L=īēāōū',
    'C=ptcqbdgmnlrhs',
    'F=ie',
    'B=ou',
    'S=ptc',
    'Z=bdg'
  ],
  rewriteRules: ['lh=lj'],
  soundChanges: [
    '[sm]//_#',
    'i/j/_V',
    'L/V/_',
    'e//Vr_#',
    'v//V_V',
    'u/o/_#',
    'gn/nh/_',
    'S/Z/V_V',
    'c/i/F_t',
    'c/u/B_t',
    'p//V_t',
    'ii/i/_',
    'e//C_rV'
  ],
  lexicon: [
    'lector',
    'doctor',
    'focus',
    'jocus',
    'districtus',
    'cīvitatem',
    'adoptare',
    'opera',
    'secundus',
    'fīliam',
    'pōntem'
  ],
  outputFormat: 'oo',
  showDiff: true,
  showChanges: true,
  ruleReport: true,
  rewriteOutput: true,
  results: undefined
}
