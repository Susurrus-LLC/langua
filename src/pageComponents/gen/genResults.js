import React from 'react'
import PropTypes from 'prop-types'

import Results from '../../components/results'

import sharedStyles from '../../components/results/sharedResults.module.sass'

const GenResults = () => <Results />

GenResults.propTypes = {
  styles: PropTypes.string
}

export default GenResults
