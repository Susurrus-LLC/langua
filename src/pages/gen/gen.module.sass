@import ../../styles/variables.module

.duplicate
  outline: 1px solid $red

.subps-row, .adding-row, .pattern-input
  width: 100%

.subps-row, .adding-row
  display: flex
  flex-direction: row
  align-items: baseline
  align-content: space-between

.subps-row
  justify-content: space-between

.adding-row, .pattern-input
  margin: 1px 0
  padding: 1px

.adding-row
  justify-content: flex-end

.subpattern-input
  flex-grow: 1

.words-input
  width: $ms6 * 1rem

.out-text
  white-space: pre-wrap
