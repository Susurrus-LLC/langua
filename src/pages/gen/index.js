import React, { useState } from 'react'
import { graphql, useStaticQuery } from 'gatsby'

import Tool from '../../components/tool'

import GenForm from '../../pageComponents/gen/genForm'
import GenResults from '../../pageComponents/gen/genResults'
import genService from '../../services/genService'

import styles from './gen.module.sass'

export const PureGen = ({ data }) => {
  const toolInfo = data?.site?.siteMetadata?.toolInfo?.gen

  // State

  const [subpatterns, setSubpatterns] = useState(
    genService.getData().subpatterns
  )
  const [pattern, setPattern] = useState(genService.getData().pattern)
  const [words, setWords] = useState(genService.getData().words)
  const [newline, setNewline] = useState(genService.getData().newline)
  const [filterdupes, setFilterdupes] = useState(
    genService.getData().filterdupes
  )
  const [status, setStatus] = useState(['ok'])
  const [results, setResults] = useState([''])
  const [stats, setStats] = useState({
    words: 0,
    maxWords: 0,
    filtered: 0,
    remaining: 0
  })

  // Functions

  const onChangeInput = e => {
    const data = {
      subpatterns,
      pattern,
      words,
      newline,
      filterdupes
    }

    if (e.target.tagName === 'BUTTON') {
      e.preventDefault()
      if (e.target.type === 'submit') {
        // Generate the output
        const response = genService.generate(data)
        setStatus(response.status)
        setResults(response.results)
        setStats(response.stats)
        // Save the current state to storage
        genService.setStorage(data)
      } else if (e.target.type === 'button') {
        if (e.target.id === 'save') {
          // Save the current state to storage and generate a file
          genService.save(data)
        } else if (e.target.id === 'add') {
          // When the add button is clicked, add a blank Subpattern to state
          setSubpatterns(genService.add(data).subpatterns)
        } else if (e.target.id.slice(0, 1) === 'c') {
          // When a Subpattern is cleared, delete the corresponding Subpattern from state
          const id = e.target.id.slice(1)
          setSubpatterns(genService.clear(id, data).subpatterns)
        }
      }
    } else if (e.target.type === 'file') {
      e.preventDefault()
      // Open a file and parse it to restore a saved state
      const file = e.target.files[0]
      const updateState = response => {
        if (response) {
          // Only change state if the file was successfully opened
          setSubpatterns(response.subpatterns)
          setPattern(response.pattern)
          setWords(response.words)
          setNewline(response.newline)
          setFilterdupes(response.filterdupes)
        }
      }
      genService.open(file, updateState)
    } else if (e.target.type === 'checkbox') {
      const checked = e.target.checked
      if (e.target.id === 'newline') {
        // If the selection for new lines is changed, store that change in state
        setNewline(genService.changeNewline(checked, data).newline)
      } else if (e.target.id === 'filterdupes') {
        // If the selection for filtering duplicates is changed, store that change in state
        setFilterdupes(genService.changeDupes(checked, data).filterdupes)
      }
    } else {
      const val = e.target.value
      if (e.target.id === 'pattern') {
        // When the pattern is changed, store the change in state
        setPattern(genService.changePattern(val, data).pattern)
      } else if (e.target.id === 'words') {
        // When the number of desired words is changed, store that change in state
        const response = genService.wordNumChange(val, data)
        if (response) {
          // Only update state if there's a change
          setWords(response.words)
        }
      } else {
        const id = e.target.id.slice(1)
        const which = e.target.id.slice(0, 1)
        if (which === 'v') {
          // When a Subpattern variable is changed, store that change in state
          setSubpatterns(genService.changeSelect(id, val, data).subpatterns)
        } else if (which === 'p') {
          // When a Subpattern is changed, store that change in state
          setSubpatterns(genService.changeSubpattern(id, val, data).subpatterns)
        }
      }
    }
  }

  return (
    <Tool toolInfo={toolInfo}>
      <GenForm
        styles={styles}
        data={{
          subpatterns,
          pattern,
          words,
          newline,
          filterdupes
        }}
        change={onChangeInput}
      />
      <GenResults
        styles={styles}
        newline={newline}
        filterDupes={filterdupes}
        status={status}
        results={results}
        stats={stats}
      />
    </Tool>
  )
}

const Gen = props => {
  const data = useStaticQuery(graphql`
    query GenToolQuery {
      site {
        siteMetadata {
          toolInfo {
            gen {
              title
              link
              description
            }
          }
        }
      }
    }
  `)

  return <PureGen {...props} data={data} />
}

export default Gen
