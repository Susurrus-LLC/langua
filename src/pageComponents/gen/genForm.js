import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'

import Form from '../../components/form'
import Button from '../../components/button'
import Control from '../../components/control'
import ControlSide from '../../componnets/control/side'
import ControlPiece from '../../components/control/piece'

import sharedStyles from '../../components/form/sharedForm.module.sass'

import { vars } from '../../data/gen'

const GenForm = () => {
  const filePicker = React.createRef()

  const selectOptions = vars.map((variab, index) => (
    <option key={index} value={variab}>
      {variab}
    </option>
  ))

  const subpatternRows = data.subpatterns.map((subpattern, index) => (
    <div key={index} className={styles.subpsRow}>
      <select
        id={`v${index}`}
        name={`v${index}`}
        value={subpattern.selected}
        onChange={change}
        className={styles.select}
        aria-label={`Variable name for subpattern ${index}`}
      >
        {selectOptions}
      </select>
      <input
        type='text'
        id={`p${index}`}
        name={`p${index}`}
        className={styles.subpatternInput}
        value={subpattern.subpattern}
        onChange={change}
        aria-label={`Subpattern ${index}`}
      />
      <div className={styles.clearButton}>
        <Button
          id={`c${index}`}
          onCLick={change}
          ver='danger'
          small
          type='button'
          aria-label={`Delete subpattern ${index}`}
        >
          Delete
        </Button>
      </div>
    </div>
  ))

  return <Form name='gen-form' />
}

GenForm.propTypes = {
  styles: PropTypes.string
}

export default GenForm
