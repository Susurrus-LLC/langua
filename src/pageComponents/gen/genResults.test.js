import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import GenResults from './genResults'
import { defData } from '../../data/gen'

describe('GenResults', () => {
  const styles = {
    outText: 'outText',
    output: 'output'
  }

  const stats = {
    words: 0,
    maxWords: 0,
    filtered: 0,
    remaining: 0
  }

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <GenResults
        styles={styles}
        newline={defData.newline}
        filterDupes={defData.filterdupes}
        status={['ok']}
        results={['']}
        stats={stats}
      />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('renders accurately with various props', () => {
    let tree = renderer
      .create(
        <GenResults
          styles={styles}
          newline={defData.newline}
          filterDupes={defData.filterdupes}
          status={['ok']}
          results={['']}
          stats={stats}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()

    tree = renderer
      .create(
        <GenResults
          styles={styles}
          newline={defData.newline}
          filterDupes={defData.filterdupes}
          status={['ok']}
          results={['a', 'sa', 'na', 'asa', 'ana']}
          stats={stats}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()

    tree = renderer
      .create(
        <GenResults
          styles={styles}
          newline={defData.newline}
          filterDupes={defData.filterdupes}
          status={['The Subpatterns contain a circular reference.']}
          results={['']}
          stats={stats}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
