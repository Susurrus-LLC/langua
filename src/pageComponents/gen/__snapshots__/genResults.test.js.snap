// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenResults renders accurately with various props 1`] = `
<div
  className="results"
>
  <div
    className="output"
  >
    <p
      className="outText"
    >
      
    </p>
  </div>
  <div
    className="stats"
  >
    <p
      className="statsText"
    >
      words: 0; maximum different words: 0
    </p>
  </div>
</div>
`;

exports[`GenResults renders accurately with various props 2`] = `
<div
  className="results"
>
  <div
    className="output"
  >
    <p
      className="outText"
    >
      a sa na asa ana
    </p>
  </div>
  <div
    className="stats"
  >
    <p
      className="statsText"
    >
      words: 0; maximum different words: 0
    </p>
  </div>
</div>
`;

exports[`GenResults renders accurately with various props 3`] = `
<div
  className="results"
>
  <div
    className="output"
  >
    <p
      className="error"
    >
      The Subpatterns contain a circular reference.
    </p>
    <p
      className="outText"
    >
      
    </p>
  </div>
  <div
    className="stats"
  >
    <p
      className="statsText"
    >
      words: 0; maximum different words: 0
    </p>
  </div>
</div>
`;
